File: mcp-code-intelligence-python/src/mcp_code_intel/analyzers/graph_analysis/utils/tarjan_scc.py
"""KSA-163: Tarjan's SCC algorithm."""
from __future__ import annotations
from .. import AdjacencyList


class TarjanSCC:
    def find_sccs(self, graph: AdjacencyList) -> list[list[int]]:
        self._index = 0
        self._stack: list[int] = []
        self._indices: dict[int, int] = {}
        self._lowlinks: dict[int, int] = {}
        self._on_stack: set[int] = set()
        self._sccs: list[list[int]] = []

        for node in graph:
            if node not in self._indices:
                self._strong_connect(node, graph)
        return [scc for scc in self._sccs if len(scc) > 1]

    def _strong_connect(self, v: int, graph: AdjacencyList) -> None:
        self._indices[v] = self._index
        self._lowlinks[v] = self._index
        self._index += 1
        self._stack.append(v)
        self._on_stack.add(v)

        for w in graph.get(v, []):
            if w not in self._indices:
                self._strong_connect(w, graph)
                self._lowlinks[v] = min(self._lowlinks[v], self._lowlinks[w])
            elif w in self._on_stack:
                self._lowlinks[v] = min(self._lowlinks[v], self._indices[w])

        if self._lowlinks[v] == self._indices[v]:
            scc: list[int] = []
            while True:
                w = self._stack.pop()
                self._on_stack.discard(w)
                scc.append(w)
                if w == v:
                    break
            self._sccs.append(scc)

File: backend/src/engine/analyzers/graph-analysis/utils/TarjanSCC.ts
/**
 * KSA-163: Tarjan's Strongly Connected Components algorithm.
 * Finds all cycles in a directed graph.
 */

import type { AdjacencyList } from '../types.js';

export class TarjanSCC {
  private index = 0;
  private stack: number[] = [];
  private indices: Map<number, number> = new Map();
  private lowlinks: Map<number, number> = new Map();
  private onStack: Set<number> = new Set();
  private sccs: number[][] = [];

  /** Find all strongly connected components with size > 1 (cycles). */
  findSCCs(graph: AdjacencyList): number[][] {
    this.reset();
    for (const node of graph.keys()) {
      if (!this.indices.has(node)) {
        this.strongConnect(node, graph);
      }
    }
    // Only return SCCs with more than 1 node (actual cycles)
    return this.sccs.filter(scc => scc.length > 1);
  }

  private strongConnect(v: number, graph: AdjacencyList): void {
    this.indices.set(v, this.index);
    this.lowlinks.set(v, this.index);
    this.index++;
    this.stack.push(v);
    this.onStack.add(v);

    const neighbors = graph.get(v) || [];
    for (const w of neighbors) {
      if (!this.indices.has(w)) {
        this.strongConnect(w, graph);
        this.lowlinks.set(v, Math.min(this.lowlinks.get(v)!, this.lowlinks.get(w)!));
      } else if (this.onStack.has(w)) {
        this.lowlinks.set(v, Math.min(this.lowlinks.get(v)!, this.indices.get(w)!));
      }
    }

    if (this.lowlinks.get(v) === this.indices.get(v)) {
      const scc: number[] = [];
      let w: number;
      do {
        w = this.stack.pop()!;
        this.onStack.delete(w);
        scc.push(w);
      } while (w !== v);
      this.sccs.push(scc);
    }
  }

  private reset(): void {
    this.index = 0;
    this.stack = [];
    this.indices = new Map();
    this.lowlinks = new Map();
    this.onStack = new Set();
    this.sccs = [];
  }
}
